packages/svelte/src/internal/client/runtime.js

```diff
@@ -1307,6 +1405,132 @@ export function derived(init) {
+}
+
+/**
+ * @param {import("./types.js").SourceSignal<unknown>} signal
+ */
+function is_last_current_dependency(signal) {
+	if (current_dependencies !== null) {
+		return current_dependencies[current_dependencies.length - 1] === signal;
+	} else if (current_consumer !== null && current_dependencies_index > 0) {
+		return current_consumer.d?.[current_dependencies_index - 1] === signal;
+	}
+	return false;
+}
+
+/**
+ * @template V
+ * @param {import("./types.js").ComputationSignal<V>} signal
+ * @param {V} value
+ * @param {ProxyHandler<any>} handler
+ * @param {(string | symbol)[]} path
+ * @returns {V}
+ */
+function proxify_object(signal, value, handler, path) {
+	const keys = new Set(Reflect.ownKeys(/** @type {object} */ (value)));
+	const proxy = new Proxy(value, handler);
+	const derived_value = /** @type {import('./types.js').DerivedSignalValue<V>} */ (signal.v);
+	let proxied_objects = derived_value.o;
+	if (proxied_objects === null) {
+		derived_value.o = proxied_objects = new Map();
+	}
+	proxied_objects.set(value, {
+		x: proxy,
+		k: keys,
+		p: path
+	});
+	return proxy;
+}
+
+/**
+ * @param {any} value
+ */
+function is_primitive_or_function_or_state_object(value) {
+	const type = typeof value;
+	return (
+		value == null ||
+		type === 'function' ||
+		type === 'string' ||
+		type === 'number' ||
+		type === 'boolean' ||
+		type === 'symbol' ||
+		type === 'bigint' ||
+		STATE_SYMBOL in value
+	);
+}
+
+/**
+ * @template V
+ * @param {import("./types.js").ComputationSignal<V>} signal
+ * @param {V} derived_value
+ * @returns {V}
+ */
+function create_derived_proxy(signal, derived_value) {
+	const handler = {
+		/**
+		 * @param {any} target
+		 * @param {string | symbol} prop
+		 * @param {any} receiver
+		 */
+		get(target, prop, receiver) {
+			const value = Reflect.get(target, prop, receiver);
+			const derived_value = /** @type {import('./types.js').DerivedSignalValue<V>} */ (signal.v);
+			const proxied_objects = /** @type {any} */ (derived_value.o);
+			const proxied_object = proxied_objects.get(target);
+			if (proxied_object === undefined) {
+				return value;
+			}
+			const { k: keys, p: path } = proxied_object;
+
+			// Only apply the proxing when we really need it. Otherwise avoid it entirely.
+			if (
+				effect_active_and_not_render_effect() &&
+				keys.has(prop) &&
+				is_last_current_dependency(signal)
+			) {
+				// Avoid generating the new_path array until we know it needed to avoid CPU cycles.
+				let new_path;
+				// We only track paths for primitives, functions or state objects, to avoid tracking objects
+				// that likely change often between derived calls. We might need to tweak this heuristic if
+				// it doesn't feel right.
+				if (is_primitive_or_function_or_state_object(value)) {
+					new_path = [...path, prop];
+					if (current_derived_proxy_property !== null) {
+						capture_fine_grain_derived_property();
+					} else {
+						current_derived_proxy_property = { s: signal, p: new_path };
```

Comment:
Why is this only set in the else branch? If I do `a.b.c.d` wouldn't that mean `b` sets `current_derived_proxy_property`, `c` reads it but resets it to `null`, and then `d` can't continue picking up the signal and path? Or asked differently, if this is correct, why would `d` want `current_derived_proxy_property` to be `null`?

Reply:
We only come to either code path when reading a property that is `is_primitive_or_function_or_state_object`. I guess I'd need a test to understand what you mean. The reason we need this is because we might have a case like

`a.b[foo].c` where `foo` is another signal.

Reply:
Imagine you have `a.b.c.d` then you would get into this code path three times, for `b`, `c` and `d`. Why is it correct that `current_derived_proxy_property` is set to something for `b`, then set to `null` for `c` (because of the if-condition seeing that `current_derived_proxy_property` is set), then set to something again for `d`?

Reply:
I'm having trouble understanding what you mean still. The property shape is fixed, it can't change dynamically. So if you have `a.b.c.d` then `b` is always an object and thus can never be `null`. If the derived itself changes, then we create a whole new proxy for that new shape and throw the old one away.